src/combinatorics/simple_graph/trails.lean

```diff
@@ -0,0 +1,168 @@
+/-
+Copyright (c) 2022 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import combinatorics.simple_graph.connectivity
+import data.nat.parity
+
+/-!
+
+# Trails and Eulerian trails
+
+This module contains additional theory about trails, including Eulerian trails (also known
+as Eulerian circuits).
+
+## Main definitions
+
+* `simple_graph.walk.is_eulerian` is the predicate that a trail is an Eulerian trail.
+* `simple_graph.walk.is_trail.even_countp_edges_iff` gives a condition on the number of edges
+  in a trail that can be incident to a given vertex.
+* `simple_graph.walk.is_eulerian.even_degree_iff` gives a condition on the degrees of vertices
+  when there exists an Eulerian trail.
+* `simple_graph.walk.is_eulerian.card_odd_degree` gives the possible numbers of odd-degree
+  vertices when there exists an Eulerian trail.
+
+## Todo
+
+* Prove that there exists an Eulerian trail when the conclusion to
+  `simple_graph.walk.is_eulerian.card_odd_degree` holds.
+
+## Tags
+
+Eulerian trails
+
+-/
+
+namespace simple_graph
+variables {V : Type*} {G : simple_graph V}
+
+namespace walk
+
+/-- The edges of a trail as a finset, since each edge in a trail appears exactly once. -/
+@[reducible] def is_trail.edges_finset {u v : V} {p : G.walk u v}
+  (h : p.is_trail) : finset (sym2 V) :=
+⟨p.edges, h.edges_nodup⟩
+
+variables [decidable_eq V]
+
+lemma is_trail.even_countp_edges_iff {u v : V} {p : G.walk u v} (ht : p.is_trail) (x : V) :
+  even (p.edges.countp (λ e, x ∈ e)) ↔ (u ≠ v → x ≠ u ∧ x ≠ v) :=
+begin
+  induction p with u u v w huv p ih,
+  { simp, },
+  { rw [cons_is_trail_iff] at ht,
+    specialize ih ht.1,
+    simp only [list.countp_cons, ne.def, edges_cons, sym2.mem_iff],
+    split_ifs with h,
+    { obtain (rfl | rfl) := h,
+      { rw [nat.even_add_one, ih],
+        simp only [huv.ne, imp_false, ne.def, not_false_iff, true_and, not_forall, not_not,
+          exists_prop, eq_self_iff_true, not_true, false_and, and_iff_right_iff_imp],
+        rintro rfl rfl,
+        exact G.loopless _ huv, },
+      { rw [nat.even_add_one, ih, ← not_iff_not],
+        simp only [huv.ne.symm, ne.def, eq_self_iff_true, not_true, false_and, not_forall,
+          not_false_iff, exists_prop, and_true, not_not, true_and, iff_and_self],
+        rintro rfl,
+        exact huv.ne, } },
+    { rw not_or_distrib at h,
+      simp only [h.1, h.2, not_false_iff, true_and, add_zero, ne.def] at ih ⊢,
+      rw ih,
+      split;
+      { rintro h' h'' rfl,
+        simp only [imp_false, eq_self_iff_true, not_true, not_not] at h',
+        cases h',
+        simpa using h } } },
+end
+
+/-- An *Eulerian trail* (also known as an "Eulerian path") is a walk
+`p` that visits every edge exactly once.  The lemma `simple_graph.walk.is_eulerian.is_trail` shows
```

Comment:
```suggestion
`p` that visits every edge exactly once. The lemma `simple_graph.walk.is_eulerian.is_trail` shows
```

Reply:
I appreciate the suggestion, but both one and two spaces after a period are accepted styles (or at least it was the APA style until recently), and the mathlib style guides don't give an opinion.